km/km_filesys.c

```diff
@@ -1649,7 +1670,27 @@ uint64_t km_fs_accept(km_vcpu_t* vcpu, int sockfd, struct sockaddr* addr, sockle
    if (ret != 0) {
       return ret;
    }
-   int hostfd = __syscall_3(SYS_accept, host_sockfd, (uintptr_t)addr, (uintptr_t)addrlen);
+   int hostfd;
+   if (snap_listen_sock == sockfd) {
+      hostfd = dup(snap_conn_sock);
+      close(snap_conn_sock);
+      snap_conn_sock = -1;
+      if (addrlen != NULL) {
+         *addrlen = sc_conn.addrlen;
+      }
+      if (addr != NULL) {
+         memcpy(addr, sc_conn.addr, sc_conn.addrlen);
+      }
+      snap_listen_sock = -1;
+   } else {
+      if ((hostfd = __syscall_4(SYS_accept4, host_sockfd, (uintptr_t)addr, (uintptr_t)addrlen, fl)) < 0) {
+         return hostfd;
+      }
+   }
+   if (fl != 0) {
+      int tmp = fcntl(hostfd, F_GETFL, NULL);
+      fcntl(hostfd, F_SETFL, tmp | (fl & (SOCK_NONBLOCK | SOCK_CLOEXEC)));
+   }
```

Comment:
This should only be done if snap_listen_sock == sockfd.
For the normal case, the accept system call will have done it.